Plutarch/Bool.hs

```diff
@@ -117,3 +146,42 @@ por = phoistAcyclic $ plam $ \x y -> pif' # x # (phoistAcyclic $ pdelay $ pcon P
 -- | Hoisted, Plutarch level, strictly evaluated boolean or function.
 por' :: Term s (PBool :--> PBool :--> PBool)
 por' = phoistAcyclic $ plam $ \x y -> pif' # x # (pcon PTrue) # y
+
+gpeq ::
+  forall a s code pcode.
+  ( code ~ Code (a s)
+  , pcode ~ ToPType2 code
+  , All2 PEq pcode
+  ) =>
+  SOP (Term s) pcode ->
+  SOP (Term s) pcode ->
+  Term s PBool
+gpeq (SOP c1) (SOP c2) =
+  ccompare_NS (Proxy @(All PEq)) (pcon PFalse) eqProd (pcon PFalse) c1 c2
+  where
+    eqProd :: All PEq xs => NP (Term s) xs -> NP (Term s) xs -> Term s PBool
+    eqProd p1 p2 =
+      foldl' (#&&) (pcon PTrue) $
+        hcollapse $ hcliftA2 (Proxy :: Proxy PEq) eqTerm p1 p2
+      where
+        eqTerm :: forall a. PEq a => Term s a -> Term s a -> K (Term s PBool) a
+        eqTerm a b =
+          K $ a #== b
```

Comment:
you may have already found this out but be careful of polymorphism in the case of recursion! This will invoke the [infinite AST problem] if the data type is recursive! (e.g `PList`)

Reply:
I hadn't thought of it! But yes, this is indeed a problem: https://hercules-ci.com/github/Plutonomicon/plutarch/jobs/998

Reply:
Fixing #19 will be beyond the scope of this PR. Generics cannot help here. 

I hand-wrote a PEq instance for PPair:

```haskell
instance (PEq a, PEq b) => PEq (PPair a b) where
  p1 #== p2 = go # p1 # p2 
    where 
      go = phoistAcyclic $ trace "pair:plam" $ plam $ \x' y' ->
        pmatch x' $ \(PPair a b) -> pmatch y' $ \(PPair a' b') -> 
          a #== a' #&& b #== b'
```

And then defined a tree type in terms of it:

```haskell
data PTree a s
  = PLeaf (Term s a)
  | PBranch (Term s (PPair (PTree a) (PTree a)))
  deriving stock (GHC.Generic)
  deriving anyclass (Generic, PlutusType)
```

PTree's PEq instance:

```haskell
instance PEq a => PEq (PTree a) where 
  t1 #== t2 = go # t1 # t2
    where 
      go = phoistAcyclic $ trace "plam"  $ 
        plam $ \x' y' -> 
          pmatch x' $ \x -> pmatch y' $ \y -> 
            case (x, y) of 
              (PLeaf a, PLeaf b) -> 
                a #== b
              (PBranch a, PBranch b) ->
                a #== b
              _ -> 
                pcon PFalse
```

No generics is involved; but comparing two tree values causes a mutual recursive loop in the compiler, as evidenced by the flurry of interleaved `plam` and `ppair:plam` traces messages.

<img width="492" alt="image" src="https://user-images.githubusercontent.com/3998/156017237-e6f6255d-1163-4b3e-ba15-86b12ec30601.png">

Reply:
No need to fix #19. This is more of an issue with the derivation itself. We *do* have to take care of it before allowing automatic `PEq` deriving at all since this is a big trap.

Fortunately, this is the part that I was experimenting on when working on `PEq` generic derivation myself. I thought of dispatching a call to either `self` (bound by a pfix #$ plam $ \self ... ->`) or `#==` depending on the outer type and inner field type. This has certain drawbacks that I haven't been able to solve. I originally wanted some way for the compiler to optimize out `pfix`s if the `self` argument was unused - but that is not currently possible.

Reply:
Dispatching a call to `self` would work for recursive types (and I'm curious how you managed to do this generically), but not for mutually recursive types like `PTree` above. 

I was thinking a compiler-level automatic `pfix` would be the solution for both these cases, including automatic use of Y* combinator as appropriate.

Reply:
Interesting case to consider
```haskell
data A = A (Term s B) (Term s B) | ANil
data B = B (Term s A) (Term s A) | BNil
```

Reply:
The equality functions for both would refer to each other, and hence be mutually recursive. This is a pretty big problem. Not sure how we can do this automatically in a good way. Perhaps we can define "mutually recursive groups" and automatically convert it into one `pfix` somehow such that for functions `fs : List Type`, we create `f : (n : Fin (length fs)) -> index fs n`.

Reply:
We probably need to do this with #260 and #259 somehow

Reply:
Can letrec do this? Or did I completely misunderstand letrec?

Reply:
`letrec` does this in an explicit way, we need implicit mutual recursion essentially. @srid we should just assume non recursive data types right now. A constructor should not contain the type being derived for for deriving to work right now. Fixing this would require a *lot* of work.

Reply:
I am finishing @TotallyNotChase 's original idea of always using `pfix` in the instance implementation, and thereby calling `self` for when checking equality of the inner type ... but it does come with the cost of slightly increased cpu/mem numbers (due to pfix being used regardless of data type being recursive). 

Do we keep that approach ... or just have generic deriving unconditionally fail on types like `PList` (where one of the constructors contains the type being derived for), if not for types like `A`/`B` above?